src/Workspaces/Core/Portable/Shared/Extensions/ITypeSymbolExtensions.ReplaceTypeParameterBasedOnTypeConstraintVisitor.cs

```diff
@@ -98,13 +107,13 @@ public override ITypeSymbol VisitTypeParameter(ITypeParameterSymbol symbol)
                         if (symbol.ConstraintTypes.All(t => t is INamedTypeSymbol))
                         {
                             var immutableProjects = _solution.Projects.ToImmutableHashSet();
-                            var derivedImplementedTypesOfEachConstraintType = symbol.ConstraintTypes.Select(ct =>
+                            var derivedImplementedTypesOfEachConstraintType = (await Task.WhenAll(symbol.ConstraintTypes.Select(async ct =>
```

Comment:
Alternative approach would be to push the async operation even higher up - store `symbol.ConstraintTypes` here to some result builder list and and perform the processing that's in the lambda outside of the visitor - the visitor would return the builder and then the caller would run the SymbolFinder on the symbols stored in the builder.

Reply:
This seems like a lot of work to refactor the behavior, and I'm not sure the result would be as easy to read. Is there a particular advantage you'd be hoping to realize from such an effort?

Reply:
IDK how much work that is, so maybe not. But in general I think it seems cleaner to separate the two steps - one step is analyze the symbols (visitor) and then act on the results in a separate step.

Reply:
i'm ok with teh new approach myself.